libraries/standard/mqtt/src/mqtt.c

```diff
@@ -99,6 +101,427 @@ static int32_t sendPacket( MQTTContext_t * pContext,
 
 /*-----------------------------------------------------------*/
 
+/**
+ * @brief Convert a byte indicating a publish ack type to an #MQTTPubAckType_t.
+ *
+ * @param[in] packetType First byte of fixed header.
+ *
+ * @return Type of ack.
+ */
+static MQTTPubAckType_t getAckFromPacketType( uint8_t packetType )
+{
+    MQTTPubAckType_t ackType;
+    switch( packetType )
+    {
+        case MQTT_PACKET_TYPE_PUBACK:
+            ackType = MQTTPuback;
+            break;
+        case MQTT_PACKET_TYPE_PUBREC:
+            ackType = MQTTPubrec;
+            break;
+        case MQTT_PACKET_TYPE_PUBREL:
+            ackType = MQTTPubrel;
+            break;
+        case MQTT_PACKET_TYPE_PUBCOMP:
+            ackType = MQTTPubcomp;
+            break;
+        default:
+            /* Not an ack. */
+            assert( 0 );
+            break;
+    }
+    return ackType;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Dump a packet from the transport interface.
+ *
+ * @param[in] PContext MQTT Connection context.
+ * @param[in] remainingLength Remaining length of the packet to dump.
+ * @param[in] bytesAlreadyReceived Bytes previously read from the transport.
+ *
+ * @return #MQTTRecvFailed or #MQTTNoDataAvailable.
+ */
+static MQTTStatus_t dumpPacket( MQTTContext_t * const pContext,
+                                size_t remainingLength,
+                                int32_t bytesAlreadyReceived )
+{
+    MQTTStatus_t status = MQTTRecvFailed;
+    int32_t totalBytesReceived = bytesAlreadyReceived, bytesReceived = 0;
+    assert( pContext != NULL );
+    size_t bytesToReceive = pContext->networkBuffer.size;
+
+    while( totalBytesReceived < ( int32_t ) remainingLength )
+    {
+        /* Update number of bytes to receive. */
+        if( ( remainingLength - ( size_t ) totalBytesReceived ) < bytesToReceive )
+        {
+            bytesToReceive = remainingLength - ( size_t ) totalBytesReceived;
+        }
+
+        bytesReceived = pContext->transportInterface.recv( pContext->transportInterface.networkContext,
+                                                           pContext->networkBuffer.pBuffer,
+                                                           bytesToReceive );
+        if( bytesReceived != ( int32_t ) bytesToReceive )
+        {
+            /* Partial receive occurred while trying to dump packet. */
+            IotLogErrorWithArgs( "Partial receive while dumping packet."
+                                 "Received bytes=%d, Expected bytes=%u",
+                                 bytesReceived,
+                                 bytesToReceive );
+            status = MQTTRecvFailed;
+            break;
+        }
+
+        /* Update number of bytes to receive. */
+        totalBytesReceived += bytesReceived;
+    }
+
+    if( totalBytesReceived == ( int32_t ) remainingLength )
+    {
+        IotLogErrorWithArgs( "Dumped packet. Dumped bytes=%d",
```

Comment:
Should this be a `Warn` or `Debug` level message instead of an error? It seems like as the intention of this function is to discard packets, then this branch should be the happy-path for this function

Reply:
No, although the intention of the function is to discard packets, this function should not be called in ideal execution of the process loop. It's mean to handle the error case when the remaining length exceeds the buffer length, and then continue execution